sway-core/src/ir_generation/const_eval.rs

```diff
@@ -1149,6 +1324,81 @@ fn const_eval_intrinsic(
                 span: intrinsic.span.clone(),
             })
         }
+        Intrinsic::EncodeBufferEmpty => Ok(Some(to_encode_buffer(lookup, vec![], 0))),
+        Intrinsic::EncodeBufferAppend => {
+            assert!(args.len() == 2);
+
+            let (slice, mut len) = as_encode_buffer(&args[0]).unwrap();
+            let mut bytes = slice.clone();
+
+            use ConstantValue::*;
+            match &args[1].value {
```

Comment:
Similar question for `()`. I understand that we have `impl AbiEncode for ()` and that `()` will always be properly encoded when being embedded. Also `().abi_encode(buffer)`. But `__encode_buffer_append` is just an available intrinsic. One can use it by calling `__encode_buffer_append(buffer, ())` and it's not illegal. Same with `bool` as discussed above. I don't see any cost involved in supporting these two types, we avoid general error that expression cannot be evaluated to const, and in general, why to have a special unsupported cases if there is no good reason to.

Reply:
The problem with supporting `()` is that to make it "complete", it would also need to support any tuple. And we would need to expand to any array, given how "similar" they are.

And my idea was to make it work only with primitive data types. 

So for example both `__encode_buffer_append(buffer, ())` and `__encode_buffer_append(buffer, MyStruct {})` fails as no const. Which I think makes sense.

Reply:
Yes makes perfectly sense, also the example comparing `()` with `EmptyStruct {}` definitely.